### PR TITLE
Simplify/improve vertical datum code. Prevent datum mixing in DSS 6.

### PR DESCRIPTION
**Pull request by perrymanmd**

### CODE REVIEW

File: heclib/heclib_c/src/Utilities/verticalDatum.c [hunk 1]
```diff
@@ -148,6 +152,20 @@ void getoffset_(
     free(cOffsetUnit);
     free(cDataUnit);
 }
+int validelevunits_(
```
Comment (MikeNeilson) [inline heclib/heclib_c/src/Utilities/verticalDatum.c:155]: it's visually hard to tell where this function actually starts.

File: heclib/heclib_c/src/Utilities/verticalDatum.c [hunk 2]
```diff
@@ -761,21 +778,24 @@ char *validateXmlStructure(const char *xml) {
 //
 // See verticalDatum.h for documentation
 //
-char *stringToVerticalDatumInfo(verticalDatumInfo *vdi, const char *inputStr) {
-    char  *errmsg = NULL;
+void initializeVerticalDatumInfo(verticalDatumInfo* vdi) {
+    memset(vdi, 0, sizeof(*vdi));
```
Comment (MikeNeilson) [inline heclib/heclib_c/src/Utilities/verticalDatum.c:838]: Needs null check.

File: heclib/heclib_c/src/Utilities/verticalDatum.c [hunk 3]
```diff
@@ -1047,7 +1067,96 @@ verticalDatumInfo *extractVerticalDatumInfoFromUserHeader(const int *userHeader,
...
+        *offsetToNgvd29           = UNDEFINED_VERTICAL_DATUM_VALUE;
+        *offsetToNgvd29IsEstimate = TRUE;
+        *offsetToNavd88           = UNDEFINED_VERTICAL_DATUM_VALUE;
+        *offsetToNavd88IsEstimate = TRUE;
+    }
+}
+//
+// See verticalDatum.h for documentation
+//
+int	getCurrentVerticalDatum(
```
Comment (MikeNeilson) [inline heclib/heclib_c/src/Utilities/verticalDatum.c:1215]: should also have a null check.

**Review by MikeNeilson — COMMENTED** [review]
I stopped reviewing at the fortran as I'm not as familiar.

For the C, it appears that some of the places I suggested require a null check are only called from code where the object is guaranteed to exist, it should still null check anyways; who knows what the next programmer will do.

**Review by MikeNeilson — APPROVED** [review]
FYI, Stopped at the fortran again.

File: heclib/heclib_c/src/Utilities/verticalDatum.c [hunk 4]
```diff
@@ -34,26 +34,93 @@ static const unsigned char base64bytes[] = {255,255,255,255,255,255,255,255,255,
...
-
-void *_calloc(size_t num, size_t size) {
-    return _malloc(num * size);
+/**
+ * calloc with zero-initialization
+ * @param num The number of items to allocate memory for
+ * @size      The size of each item
+ * @return    The allocated memory or NULL if unable
+ */
+void* _calloc(size_t num, size_t size) {
```
[the remark this answers is not in the document]
Comment (perrymanmd) [reply heclib/heclib_c/src/Utilities/verticalDatum.c:72]: Good point. Was supposed to be temporary, but, well, you know...

**Review by MikeNeilson — APPROVED** [review]
I know they're required because of how things work, but all those casts to long long* inside the free call just make me a little sad.

**Pull request merged**